feat: Build optimistic Edit Task actions

paultsimura opened the pull request:
### Details


### Fixed Issues

$ https://github.com/Expensify/App/issues/36196
PROPOSAL: https://github.com/Expensify/App/issues/36196#issuecomment-1935181649


### Tests

Same as QA

### Offline tests

Same as QA

### QA Steps

1. Create a task
2. Go offline
3. Add a description
4. Verify the `updated the description to {value}` report action is added
5. Modify the title
6. Verify another `updated the task title to {value}` report action is added
7. Remove the description
8. Verify the `removed the description` report action is added
9. Assign somebody
10. Verify the `assigned to @user` report action is added
11. Go online
12. Refresh the page
13. Verify all the report actions remain

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/12595293/9b06ef42-b533-4c0f-a46b-165edb4f0992


</details>


<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/12595293/b1d8fb13-22d7-480f-a29a-2c6f317a7efa


</details>


<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/12595293/d9bdcde4-2a2a-4ca0-805f-83d74dd42378


</details>


<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/12595293/8ec80c09-aca2-461a-8b15-81c66e712de2


</details>


<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/12595293/e3bf2856-1768-4868-b598-48aaf181de10


</details>


<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/12595293/c909b32b-ce2d-41bd-b5a7-ce33a5a00373


</details>

paultsimura committed:
Build optimistic Edit Task actions

paultsimura commented:
@thienlnam there are some points I'd like to discuss with you.

**First** of all, the `EditTaskAssignee` and `EditTask` API calls do not return the "modified" `reportAction` in the response – it comes later from the `OpenReport`, so we need to refresh (or revisit) the report page to replace the optimistic report action with the one from the server – this should be fixed on the BE side.

**The second** is related to the mentions when changing the assignee. Currently, they come from the BE in this format:
```json
[
  {
    "type": "TEXT",
    "style": "normal",
    "text": "assigned to <mention-user>user_d (paultsimura+d@gmail.com)</mention-user>"
  }
]
```

This is incorrect in a couple of ways. First, the `<mention-user>` tag [doesn't support this format](https://expensify.slack.com/archives/C01GTK53T8Q/p1708012922231739?thread_ts=1708004466.085689&cid=C01GTK53T8Q) – the easiest way would be to use the `accountId` param, e.g.:

```js
`assigned to <mention-user accountID=${assigneeAccountID}></mention-user>`
```
The display name will be fetched automatically by the provided `accountID` by the HTML renderer.

And one last thing is the `"type": "TEXT"` doesn't support the HTML render – it will show the action as a plain text. We need to use the `"type": "COMMENT"` in order for mentions to be displayed correctly, or think of a way to allow parsing HTML in the `TEXT` actions.

thienlnam commented:
Thanks for the comments

> First of all, the EditTaskAssignee and EditTask API calls do not return the "modified" reportAction in the response – it comes later from the OpenReport, so we need to refresh (or revisit) the report page to replace the optimistic report action with the one from the server – this should be fixed on the BE side.

Which reportAction are you referring to in this case? There should be a pusher update which updates the parent reportAction data so you don't need to fetch it from the API call. Though if you're talking about the edited reportAction, I agree we might need to update it so that it does


As for the other items, I'll get a fix up for those!

paultsimura commented:
> Though if you're talking about the edited reportAction

Yes, I'm talking about them. The optimistic ones do not get replaced with the ones from BE until the page is refreshed, we just need to add one more `merge` to the BE response.

paultsimura committed:
Merge branch 'main' into fix/36196-edit-task

paultsimura committed:
DRY

paultsimura commented:
> As for the other items, I'll get a fix up for those!

Would be great if you could test this PR's optimistic flow to see if the BE should match it – so that we'll be on the same page.

As a side note: I saw you intended to use the `Assigned to "you"` mention format from the BE – I'm not sure it's something we currently support – the self-mention will render also with the `displayName`, but highlighted in green (all other mentions will be blue).

thienlnam commented:
> As a side note: I saw you intended to use the Assigned to "you" mention format from the BE – I'm not sure it's something we currently support – the self-mention will render also with the displayName, but highlighted in green (all other mentions will be blue).

The "you" wasn't going to be a mention, but I think I'll just update it to be a mention and then it will show as green - that's okay as well

paultsimura commented:
> The "you" wasn't going to be a mention

It's currently coming from BE like this, therefore I assumed the mention 🙂 
<img width="424" alt="image" src="https://github.com/Expensify/App/assets/12595293/4fc4ef3d-05c4-44f9-910e-6419b3741228">

thienlnam commented:
![Screenshot 2024-02-20 at 6 06 24 PM](https://github.com/Expensify/App/assets/30609178/280734ad-f272-467e-8419-b5f5bc69bff3)

I tested the changes with your branch, looks good - seems like the only change that is needed is for the comments to show as system messages UI

paultsimura commented:
We have +- 2 formats of system messages:

![image](https://github.com/Expensify/App/assets/12595293/35970d3e-4897-465a-a4f2-d51bf404eed7)

![image](https://github.com/Expensify/App/assets/12595293/73feeba0-ab19-452e-9fe1-c9426cdb023c)

Do you mean we would like the `TASKEDITED` messages to look like the room renames (the second screenshot)?

paultsimura committed:
Merge branch 'main' into fix/36196-edit-task

thienlnam commented:
No no, they should look like the first screenshot - I more meant that it the UI should be in the same text color from this screenshot https://github.com/Expensify/App/pull/36768#issuecomment-1955740764

Though, I  don't think I've seen a mention within a system message so this might be new

paultsimura commented:
Ah yes - that's a bit of an issue — the bold system-like formatting is supported by the actions of type `TEXT`, but these actions do not support mentions. And vice-versa, the `COMMENT` type supports mentions, but not the formatting. I can look what we can do with this formatting, but need to know if it's worth the time at this point

thienlnam commented:
Hahah of course there is something like that - I think for now it's okay to keep as is so no need to spend additional time investigating. Thanks for being willing though

paultsimura committed:
No longer hide the TASKEDITED actions

thienlnam commented:
Yes that's correct - the PRs are in review currently. I'll let you know when they're on production

paultsimura committed:
Merge branch 'main' into fix/36196-edit-task

thienlnam commented:
Changes should be on production now

paultsimura committed:
Merge branch 'main' into fix/36196-edit-task

paultsimura commented:
Thanks @thienlnam. They are, so I'm unblocked now, but this one's still missing: https://github.com/Expensify/App/pull/36768#issuecomment-1955185634

The `EditTask` API response does not update the task report's `reportActions` to include the new `TASKEDITED` action. It's NAB, but would be nice to add.

Here's the recording, the `reportActions` is updated with the server data only after we refresh the page and the `OpenReport` response comes (the task report ID is `5598782205931668`):

https://github.com/Expensify/App/assets/12595293/0395423f-8dd0-4693-a438-c67898d3ba01

paultsimura commented:
@thienlnam another BE bug: when editing the assignee, we need to add their display name to the `reportAction.message.text`.

This is what an optimistic action looks like (note the text in LHN as a result):
<img width="1512" alt="image" src="https://github.com/Expensify/App/assets/12595293/734e65d9-c431-4c21-9d67-2481751ecf7c">

And here's what it becomes after coming online and being updated from the BE:
<img width="1512" alt="image" src="https://github.com/Expensify/App/assets/12595293/b0e34923-8e5a-4c69-ad61-439b5f1a3ab3">

paultsimura commented:
One more thing @thienlnam, sorry.

In money requests, we don't count the `MODIFIEDEXPENSE` system messages as visibleChildReportActions, therefore a money request with _only such messages_ can be fully deleted (if there are other messages inside – the report won't be deleted). Also, these messages are not counted as "thread replies count".
Demo:


https://github.com/Expensify/App/assets/12595293/cbee0b28-aea7-4afa-909d-03b0a4fbbff5



Do we want to implement a similar approach for tasks? And if so – should we try implementing it in the scope of this task?

paultsimura committed:
Reformat task-related actions

paultsimura commented:
>> Also, these messages are not counted as "thread replies count".
> I'll add this to the list of BE fixes needed

This is similar to how it's done with the Money Request system messages as well as the "task completed/reopened", so I would double-check that - it seems expected to me.

This PR is ready for review code-wise, so I would recommend just checking how the branch works, and letting me know if something is misaligned with your vision. 
Btw, I managed to make these messages look like the system ones.

paultsimura commented:
@thienlnam it's me again 😄 

I've taken a closer look and IMO we need to modify the way we work with the `EditTask` API endpoint on the BE.

Currently, we pass both `title` & `description` on every task update:
https://github.com/Expensify/App/blob/22daa33e308cf6c47310414aa3c340f90f548f6b/src/libs/actions/Task.ts#L444-L451

This leads to an issue:
1. Create a task
2. Go offline
3. Update the title
4. Update the description
5. Go online

At some point after coming online, when the `updateTitle` request comes, the optimistically updated description will be reverted because in that request, we passed the description that was relevant at the moment of the request (i.e. old one). And only after the second, `editDescription` response comes, the fields will become correct. In other words, we reach only eventual consistency, but not intermediate.

We need to modify the `EditTask` endpoint to allow passing only one param of `title` and `description` so that the stale data is not returned in the merge operation of the API response.

Demo of the bug:

https://github.com/Expensify/App/assets/12595293/9df4035b-2f9e-4eac-8655-0944f5edd9eb



P.s. I tried passing only one param now, but it fails with an error if we pass only `description` without `title`.

thienlnam commented:
Thanks, I've created a tracking issue for those changes we'll need to make in the BE. Let's move forward here since all those are non-blocking for this PR

eVoloshchak commented:
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

eVoloshchak commented on src/libs/ReportUtils.ts:
```diff
@@ -3416,6 +3417,10 @@ function getCurrentUserAvatarOrDefault(): UserUtils.AvatarSource {
 }
 
+function getCurrentUserDisplayNameOrEmail(): string | undefined {
+    return allPersonalDetails?.[currentUserAccountID ?? '']?.displayName ?? currentUserEmail;
```
```suggestion
    return currentUserPersonalDetails?.displayName ?? currentUserEmail;
```

paultsimura replied:
Done, thanks ✔️

paultsimura committed:
Merge branch 'main' into fix/36196-edit-task

paultsimura committed:
Reuse functions to get current user data

eVoloshchak commented:
@paultsimura, there are some conflicts (app is moving fast)😅

paultsimura committed:
Merge branch 'main' into fix/36196-edit-task

# Conflicts:
#	src/pages/home/report/ContextMenu/ContextMenuActions.tsx

paultsimura committed:
Merge branch 'main' into fix/36196-edit-task

paultsimura commented:
@eVoloshchak resolved the conflicts ✔️

eVoloshchak commented:
Bug:

1. Create a task
2. Go offline
3. Add a description
4. Add an assignee
5. Mark as complete
6. Mark as incomplete
7. At this point all of the report actions are in the correct order
8. Go online
9. Notice "marked as complete" and "marked as incomplete" items changing position while the app refreshes

https://github.com/Expensify/App/assets/9059945/48e40e3a-8980-4af2-a919-9d8871127d81

eVoloshchak commented:
I think we should localize the actions, they aren't translated into Spanish


https://github.com/Expensify/App/assets/9059945/37ce601c-c963-417b-a87d-25bac1659a66

paultsimura commented:
> Bug:
> 
> 1. Create a task
> 2. Go offline
> 3. Add a description
> 4. Add an assignee
> 5. Mark as complete
> 6. Mark as incomplete
> 7. At this point all of the report actions are in the correct order
> 8. Go online
> 9. Notice "marked as complete" and "marked as incomplete" items changing position while the app refreshes

That's known: https://github.com/Expensify/App/pull/36768#issuecomment-1969152221

@thienlnam is tracking this on the BE side.

paultsimura commented:
> I think we should localize the actions, they aren't translated into Spanish

For this, we'll need another BE change.
@thienlnam could you please add to your tracking issue the following?

The `TASKEDITED` report actions should reflect the `MODIFIEDEXPENSE`:
1. The API endpoint must consume one field at a time (title / description);
2. The `originalMessage` must contain only the changed field (old + new value), as it is for expenses:

Updated Description:
```js
{
  "oldComment": "",
  "newComment": "Desc",
  "lastModified": "2024-03-04 23:22:47.451"
}
```

Updated Merchant:
```js
{
  "oldMerchant": "Old Merchant",
  "merchant": "New Merchant",
  "lastModified": "2024-03-04 23:28:22.705"
}
```

We need to make sure the `originalMessage` will cover the `title`, `description`, and `assignee` change.
Based on this, we'll be able to build a localizable message for each `TASKEDITED` action, similar to `MODIFIEDEXPENSE` here:

https://github.com/Expensify/App/blob/56097445d343c0b0f3b8a45189b014ef28429910/src/libs/ModifiedExpenseMessage.ts#L100-L239

paultsimura committed:
Merge branch 'main' into fix/36196-edit-task

paultsimura commented:
@thienlnam I have implemented the localizable optimistic actions. So if you haven't updated the API on the BE, I'll share the object type I'm currently expecting optimistically, please share what you think about it:

```ts
type OriginalMessageTaskEdited = {
    actionName: typeof CONST.REPORT.ACTIONS.TYPE.TASKEDITED;
    originalMessage: {
        oldTitle?: string;
        title?: string;
        oldDescription?: string;
        description?: string;
        oldAssigneeAccountID?: number;
        assigneeAccountID?: number;
    };
};
```

This way, the `originalMessage` will be as close to the `MODIFIEDEXPENSE` as possible: it will hold only `oldValue` and `value` of the field that was modified.
Besides that, storing the `oldValue` will allow displaying messages in a format like: `updated the description to "new" (previously "old")` if we want to.

paultsimura committed:
Localize taskEdited messages

thienlnam commented:
I should have clarified but let's move forward without localization in this PR - localization is not a current priority so while I've added it to the list - it may be a while until we decide to reprioritize those changes

paultsimura committed:
Revert "Localize taskEdited messages"

This reverts commit a86a2f7d2c968526ac2215f9505a4d93f971af2a.

paultsimura commented:
Got you. Then we're good to test again @eVoloshchak.

eVoloshchak reviewed (approved):
Tested this by cherry-picking the PR into the latest main, works well, approving
@paultsimura, could you resolve the conflicts please?

paultsimura committed:
Merge branch 'main' into fix/36196-edit-task

# Conflicts:
#	src/pages/home/report/ReportActionItem.tsx

thienlnam commented:
Getting some type errors cc @paultsimura 

```
Error: src/libs/actions/Task.ts(483,93): error TS2345: Argument of type 'number | null' is not assignable to parameter of type 'number'.
  Type 'null' is not assignable to type 'number'.
```

paultsimura committed:
Fix TS error

paultsimura commented:
> Getting some type errors

Should be resolved now, let's see what the CI will say...

thienlnam commented:
All set - thanks for the quick work on this!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8196345246) to staging by https://github.com/thienlnam in version: 1.4.49-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8253954132) to production by https://github.com/luacmartins in version: 1.4.50-5 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

hungvu193 commented on src/libs/OptionsListUtils.ts:
```diff
@@ -584,12 +583,8 @@ function getLastMessageTextForReport(report: OnyxEntry<Report>, lastActorDetails
-    ) {
-        lastMessageTextFromReport = lastReportAction?.message?.[0].text ?? '';
+    } else if (ReportActionUtils.isTaskAction(lastReportAction)) {
+        lastMessageTextFromReport = TaskUtils.getTaskReportActionMessage(lastReportAction).text;
```
This line will introduce the regression https://github.com/Expensify/App/issues/37976, causing the last message text to become multiple lines in the LHN. So I think in the future whenever we develop a feature that relate to text, we should also test the text overflow case.
